velox/functions/sparksql/String.h

```diff
@@ -902,4 +902,106 @@ struct TranslateFunction {
   }
 };
 
+template <typename T>
+struct ConvFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  static constexpr bool is_default_ascii_behavior = true;
+
+  static const uint64_t kMaxUnsignedInt64_ = 0xFFFFFFFFFFFFFFFF;
+  static const int kMinBase = 2;
+  static const int kMaxBase = 36;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      int32_t fromBase,
+      int32_t toBase) {
+    if (input.empty()) {
+      return false;
+    }
+
+    // Consistent with spark, only supports fromBase belonging to [2, 36]
+    // and toBase belonging to [2, 36] or [-36, -2].
+    if (fromBase < kMinBase || fromBase > kMaxBase ||
+        std::abs(toBase) < kMinBase || std::abs(toBase) > kMaxBase) {
+      return false;
+    }
+
+    // Ignore leading spaces.
+    int i = 0;
+    for (; i < input.size(); i++) {
+      if (input.data()[i] != ' ') {
+        break;
+      }
+    }
+    const bool isNegativeInput = (input.data()[i] == '-');
+    // Skip negative symbol.
+    if (isNegativeInput) {
+      ++i;
+    }
+    uint64_t unsignedValue;
+    auto fromStatus = std::from_chars(
+        input.data() + i, input.data() + input.size(), unsignedValue, fromBase);
+    if (fromStatus.ec == std::errc::invalid_argument) {
+      result.append("0");
+      return true;
+    }
+    if (fromStatus.ec == std::errc::result_out_of_range) {
+      unsignedValue = kMaxUnsignedInt64_;
+    }
+    if (unsignedValue == 0) {
+      result.append("0");
+      return true;
+    }
+
+    int64_t signedValue;
```

Comment:
perhaps, extract this logic into a helper function; it is a bit hard to reason about why signedValue is not initialized when toBase > 0 and isNegativeInput is false. Maybe it will be easier to read this code if toBase < 0 was handled explicitly:

```
if (toBase < 0) {
   signedValue = ...
   ...convert signedValue to result
} else if (isNegativeInput) {
   // Convert negative value to unsigned value.
   unsignedValue = ...
   ...convert unsignedValue to result
}
```

Reply:
When toBase > 0 and isNegativeInput is false, we use the raw unsigned value to convert. Just refactored this part with helper functions. May look a bit better now. Thanks!